### ISSUE TITLE
Broken symlinks in emacs 26 on macOS

### ISSUE THREAD

**Issue by detvdl** [abo-abo/swiper#1239]
Recently I've switched over to the emacs-26 branch, which includes a couple of bug fixes relating to window management in macOS. However, this has introduced an issue with ivy:

Symbolic links in find-file now appear as if they were directories (highlighted/bold and with a trailing slash), and upon opening them I am presented with an empty buffer that is constantly reverting (this is due to having global-auto-revert-mode set).

I have tried to replicate this bug on my Arch Linux system, compiling emacs 26 with the same flags, but it does not seem to occur there, nor on any other emacs version (emacs 25.3, for example, does not have this issue on macOS). 

A minimal configuration containing only the `ivy` package does not fix the issue either (and hopefully excludes the possibility of any specific configuration settings being at fault).

The issue also does not present itself when using the `helm` package, though that is not enough of a reason for me to switch over :smile:.

### System info
**Operating System:** macOS High Sierra (Version 10.13)
**Emacs:** GNU Emacs 26.0.90 (build 2, x86_64-apple-darwin17.0.0, 
 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; NS appkit-1561.00 Version 10.13 (Build 17A405)) of 2017-10-13

### Screenshot
<img width="1680" alt="screen shot 2017-10-15 at 22 18 14" src="https://user-images.githubusercontent.com/8858957/31588762-cf47d7b6-b1f6-11e7-9288-c388ee1a0dcf.png">

Both emacs instances share the exact same configuration except for the display-line-numbers-mode setting, which is not included in emacs < 26 yet. This should not affect the outcome, however.

If any further information is needed, I would be happy to provide it.

**Comment by abo-abo** [comment]
Try to reproduce it like this. Setup:

```sh
cd ~
mkdir temp
touch temp/foo
ln -s temp/foo temp/bar
```

Then eval:

```elisp
(let ((default-directory "~/temp"))
  (all-completions "" 'read-file-name-internal))
```

Compare the result on Mac and Arch. If it's different, it's not an Ivy issue and can be reported with <kbd>M-x</kbd> `report-emacs-bug`.  Well, theoretically Ivy issues can be reported like this as well (it's in GNU ELPA), but realistically I'm more motivated to fix the issues anyway, so it's better to report them here.

**Comment by detvdl** [comment]
I currently do not have access to my Linux machine, so I will report on that later today.

I can, however, already tell you that emacs 26.0.90 and emacs 25.3.2 give different results when the symlinks are properly set up on macOS, like this:

```sh
cd ~
mkdir temp
touch temp/foo
ln -s ~/temp/foo ~/temp/bar
```

The output then shows:
<img width="1680" alt="screen shot 2017-10-16 at 10 39 05" src="https://user-images.githubusercontent.com/8858957/31602734-43e20cce-b25e-11e7-8e3f-dbeb8d4b2daf.png">

Creating the symlinks exactly as you wrote above leaves me with an invalid symlink. Evaluating the elisp fragment with that setup gives me the same output on both emacs versions on macOS.

I will post the output on my arch machine of both emacs versions as soon as I can.

**Comment by abo-abo** [comment]
Thanks. So if I understood correctly:

1. You modified the shell setup script to have the paths start with `~`.
2. The result of the eval of the Emacs snippet is different (`bar/` vs `bar`).

I think that sufficient information to conclude that it's an OSX-specific Emacs bug, not related to Ivy.
I can guess that the eval results will be the same on Arch (they are the same for me on Ubuntu).

**Comment by detvdl** [comment]
Correct. Using your setup script, both emacs versions returned `("bar" "foo" "../" "./")` on macOS. The script did present a faulty symlink however, which would affect the outcome.

What I don't understand, is how this could be an emacs-specific issue on OS X when I have not seen this issue trying out the `helm` package's `helm-find-files` function. I find this rather odd, especially since they do not seem to provide any specific handling for symlinks in any way.

My Emacs and emacs-lisp knowledge is far from sufficient to dispute this in any way, however, so I will direct this issue to the emacs bug tracker and see what they have to say about it.
You may close this issue if you do not find it relevant to the `ivy` package anymore.

Thanks for you interest in the issue and quick response :smile:

**Comment by abo-abo** [comment]
> What I don't understand, is how this could be an emacs-specific issue on OS X when I have not seen this issue trying out the helm package's helm-find-files function.

Emacs provides many methods for obtaining a list of files in a directory; the method `ivy-read` relies on is buggy on OSX. `helm-find-files` probably uses a different method, which isn't buggy on OSX.

**Comment by detvdl** [comment]
The bug was reported to the emacs bug tracker and has been confirmed to be a read-file-name related issue. I guess this issue can be closed now.

**Comment by abo-abo** [comment]
Thanks.

**Comment by hotpxl** [comment]
Hi @detvdl , I encountered the very same bug. I wonder if you still have the link to the bug tracker? Thanks in advance.

**Comment by hotpxl** [comment]
This has been fixed by the 26.1 release today.
I used the precompiled version from https://emacsformacosx.com/
